Support Max or Min number of decimals

malidukica opened the pull request:
This is a breaking API change, since one must now use the constructor
`Decimals.Max i` instead if just `i` to specify the maximal number
of decimal digits. It introduces a `Decimals.Min i` that will not
round down but rather ensure that at least that amount of decimal digits
are displayed, and possibly more digits, if applicable.

The existing behaviour was named `Max` because it may also not
display any decimal digits, but has a max.

The new behavour was named `Min` because there is a minimum amount
of decimal digits that are always displayed and no upper limit
on how many.

Humanize will work with Max as expected, but won't do anything to the
result of `format` if the provided decimals were Min.

cuducos commented:
Hi @malidukica and @jhf. Many thanks for the PR.

By now I'll leave aside the technical details and ask: what's the use case, for that change? I have expressed in other issues/PRs in this package that it's already a concern the complexity of the `Locale`, so I'm still not sure about what gains and trade offs this technical description you offer might bring in. Can you go through it in terms of user experience of the API and use cases?

jhf commented:
We would like to display a decimal number with at least one decimal, but possibly more, if applicable, just as described by others [1].
After seeing your comment [2] we thought about it, and `Min` and `Max` seemed intuitive names for the different behaviours, where the existing behaviour corresponds to `Max` and the new behaviour corresponds to `Min`.

Our goal was to provide the functionality we wanted with a minimum of changes, adhering to the existing suggestions [1].

We did however think about the API, and it seems that the `ZeroStrategy` and the `Decimals` are intertwined.  For instance one could imagine that instead of
```
type Decimal
  | Max Int
  | Min Int

type ZeroStrategy
    = KeepZeros
    | RemoveZeros
```
We could combine them to
```
type Decimal
  | Max Int
  | Fixed Int
  | Min Int
```
Where `Max` has the current `RemoveZeros` behaviour, and where `Fixed` corresponds to `KeepZeros` and `Min` shows at least the expected number of decimals.

Furthermore, there could, from an API standpoint, be a default records with settings, that one could
use and override, unrelated to this suggestion, but that would allow an extension of the Locale with `type DecimalStrategy = Max | Min` where `Max` was set in the default locale, and `decimals` is still used as it is today.

I think the concept of a default locale, where one can override wanted behaviour makes for a simpler API,
regardless of this PR. If there was a default, then existing code, overriding the default, would work with the new default, where `{.., decimalStrategy=Max}` without any changes.

[1] https://github.com/cuducos/elm-format-number/issues/20
[2] https://github.com/cuducos/elm-format-number/issues/20#issuecomment-529513821

cuducos commented:
Great! I this description cover what I was worried about.

I do like the design using more human-readable types (`Max Int` and `Min Int` for decimals), but I worry about causing too much confusion in the overall API. I do think a better way to handle that is to simplify the API as `Min` and `Max` offers a better abstraction for the challenge of representing decimal numbers as text (`String`).

Some questions to clarify:

1. Using `Max`, can we go back to have only `format`  and drop `humanize` altogether?
2. What would be the difference (in @jhf's comment, not in the code from the PR) between `Min` and `Fixed`? What would `1.00`  and `pi` would using `Min 2` and `Fixed 2`?
3. Would `AtLeast` be a better name than `Min`? I mean, `pi` using `Min` theoretically would mean an infinite `String`, if you know what I mean.

Once more, thank you for the contribution. I do think the abstraction you have in mind is way better than the current implementation in this package. I'm just digging it down to refine a bit here and there!

jhf commented:
> I do like the design using more human-readable types (`Max Int` and `Min Int` for decimals), but I worry about causing too much confusion in the overall API. I do think a better way to handle that is to simplify the API as `Min` and `Max` offers a better abstraction for the challenge of representing decimal numbers as text (`String`).
> 

Going with your idea of API simplification, and using possibly another, also clear, naming strategy, I see two main possibilities:

```
type Decimal
  | AtMost Int
  | Fixed Int
  | AtLeast Int
```

This allows adding a future `Range Int Int` if there exists a use case for decimals with a max limit. We believed we needed that, but after reviewing our users experience we wan't to prevent excess decimals in the input field, not in the presentation, because that minimises surprise for the user. What you enter is what we display.

If the `Range Int Int` is of no concern I think one could use

```
type alias Locale = {..., decimals : Int, decimalStrategy: DecimalStrategy}

type DecimalStrategy
  | AtMost
  | Fixed
  | AtLeast
```

but then there is less help from the compiler on where the `decimalStrategy` must be handled, so I like the former approach.

> Some questions to clarify:
> 
> 1. Using `Max`, can we go back to have only `format`  and drop `humanize` altogether?
>
I think so yes, and that simplifies the library, were it not for the weird handling where I put `???`:

```
humanize usLocale RemoveZeros 10.00
--> "10"
```
`AtMost 2`


```
humanize usLocale RemoveZeros 10.10
--> "10.1"
```
`AtMost 2`


```
humanize usLocale KeepZeros 10.00
--> "10" ???
```

`AtLeast 2` would give `10.00`

```
humanize usLocale KeepZeros 10.10
--> "10.10"
```
`AtLeast 2`

> 2. What would be the difference (in @jhf's comment, not in the code from the PR) between `Min` and `Fixed`? What would `1.00`  and `pi` would using `Min 2` and `Fixed 2`?

Well the current code is a bit strange from my perspective, removing decimal zeroes even when `KeepZeroes` is used, but regardless of that:

`AtLeast 2`aka. `Min 2`
* `1` -> `1.00`
* pi -> `3.141592653589793`

`Fixed 2`
* `1` -> `1.00`
* pi -> `3.14`

> 3. Would `AtLeast` be a better name than `Min`? I mean, `pi` using `Min` theoretically would mean an infinite `String`, if you know what I mean.
> 

In my head there is no difference between `Min` and `AtLeast` with regards to the number of decimals, that is dictated by the Float support, however, the names `Min` and `Max` associates with clearly named mathematical functions, so I think `AtLeast` and `AtMost` are good and clear, as in the examples above.

> Once more, thank you for the contribution. I do think the abstraction you have in mind is way better than the current implementation in this package. I'm just digging it down to refine a bit here and there!
>
Surly there is more to rounding than meets the eye at first glance, but that seems true of most things.

I would like the `Decimals` type to be exported by a module named `Decimals`, so the API usage would look like this:

```
import FormatNumber.Decimals exposing (Decimals)
...
usLocale : Locale
usLocale =
    Locale (Decimals.AtMost 2) "," "." "−" "" "" "" "" ""
```
That makes it more readable, or use the record constructor syntax, that spells out what everything is for
```
usLocale =
    { decimals = AtMost 2
    , thousandSeparator =  ","
    , decimalSeparator =  "."
    , negativePrefix =  "−"
    , negativeSuffix =  ""
    , positivePrefix =  ""
    , positiveSuffix =  ""
    , zeroPrefix =  ""
    , zeroSuffix =  ""
    }
```

and then finally by using a default named `default` matching Elm's String.fromFloat:
```
default =
    { decimals = AtLeast 0
    , thousandSeparator =  ""
    , decimalSeparator =  "."
    , negativePrefix =  "−"
    , negativeSuffix =  ""
    , positivePrefix =  ""
    , positiveSuffix =  ""
    , zeroPrefix =  ""
    , zeroSuffix =  ""
    }


usLocale =
    { default
    | decimals = AtMost 2
    , thousandSeparator =  ","
    , decimalSeparator =  "."
    , negativePrefix =  "−"
    }
```

I didn't think it suitable to go into this kind of detail for such a small PR, but getting a nice API is almost always a worthwhile effort, and it's clearly good to think and discuss before reorganising :-)

cuducos commented:
Great news. I like to simplify the API.

What do you think about the following summary and naming suggestions?

- [ ] Implement the **`Decimals` union type** with:
  - [ ] `Max Int` that will show **up to** a certain number of decimal digits, **discarding leading zeroes if that is the case**
   - [ ] `Min Int` that will show **at least** a certain amount of decimal digits, **adding leading zeros if needed**
  - [ ] `Exact Int` that will show an **exact** amount of decimal digits, also **adding leading zeros if needed**
- [ ] Implement a `base` locale following the behavior of  Elm's `String.fromFloat` **but** using the proper _minus_ character (and not a hyphen)  for negative numbers

Some minor details on the implementation:

1. I would call the locale you suggested locale `base` (in opposition do `default`); I don't wanna imply that a certain decimal separator, for example, is the _default_ for anything, so I prefer to be softer and call the thing `base` instead of `default`.
2. I see this PR has **a lot** of changes in the expectations of the tests and I do not think this is a good idea because it might be too abrupt to change this behavior like that (even when changing the API); I would prefer to see all tests using the `Fixed Int` thing, and the current test cases would expect the same results; and, surely, we add test cases for `Max Int` and `Min Int`.

jhf commented:
I vote for your proposal of
```
type Decimal
  = Max Int
  | Min Int
  | Exact Int
```
that replaces the current `Locale.decimals`, `ZeroStrategy` and `humanize` using your described semantics.

I also like `base` as the name for the locale that matches `String.fromFloat` using unicode minus (U+2212) `−` instead of hyphen/dash `-`.

I see the call for piecemeal changes in the commits, keeping existing tests, and adding new tests.

So there will be a breaking API change, and there will one semantic difference:
`humanize usLocale KeepZeros 10.00` becomes `10` now, and after the changes it would translate to `Min 2` that formats to `10.00`.

malidukica committed:
Add new Decimals type

Min - at least the provided number of decimals. Will add zeros if number
of actual decimals is lower. Will show decimal numbers after the limit
if the decimals if they aren't equal to zero.

Max - maximally the provided number of decimals. Will remove trailing
zeros.

Exact - exact number of decimals. Will add zeros if number of actual
decimals is lower. Doesn't show the decimal numbers after limit.

malidukica committed:
Remove Humanize functionality

Adjust tests
Remove humanize tests from Stringfy, since Min/Max/Exact do not affect
the output in stringfy (only in format).

malidukica committed:
Add tests for testing the behaviour of Max and Min

malidukica committed:
Update README.md

malidukica commented:
Hi!

Here are the new changes in regards to the ongoing discussion:

1. Added Decimals union type with Min, Max and Exact
2. Removed humanize functionality
3. Adjust tests
4. Add base locale config

cuducos commented on README.md:
```diff
@@ -11,19 +11,19 @@ import FormatNumber exposing (format)
 
-It is flexible enough to deal with different number of decimals, different thousand separators, different decimal separator, and different ways to represent negative numbers — all that is possible using `Locale`s.
+It is flexible enough to deal with different number of decimals, different thousand separators, different decimal separator, and different ways to represent negative numbers — all that is possible using `Locale`s. `base` locale matches String.fromFloat using unicode minus (U+2212) − instead of hyphen/dash -.
 
```
```suggestion
It is flexible enough to deal with different number of decimals, different thousand separators, different decimal separator, and different ways to represent negative numbers — all that is possible using `Locale`s. The `base` locale matches `String.fromFloat` using unicode minus (U+2212) instead of an hyphen or dash.
```

cuducos commented on README.md:
````diff
@@ -11,19 +11,19 @@ import FormatNumber exposing (format)
 ```elm
 import FormatNumber exposing (format)
-import FormatNumber.Locales exposing (Locale, usLocale)
+import FormatNumber.Locales exposing (Locale, usLocale, Decimals(..))
````
```suggestion
import FormatNumber.Locales exposing (Decimals(..), Locale, usLocale)
```

cuducos commented on README.md:
````diff
@@ -32,21 +32,13 @@ format usLocale -pi --> "−3.14"
-humanize usLocale KeepZeros 10.00 --> "10"
-humanize usLocale KeepZeros 10.10 --> "10.10"
-```
+1. `Min Int` - will show at least a certain amount of decimal digits, adding trailing zeros if needed
````
```suggestion
* `Min Int` shows at least a certain amount of decimal digits, adding trailing zeros if needed.
```

cuducos commented on README.md:
````diff
@@ -32,21 +32,13 @@ format usLocale -pi --> "−3.14"
-humanize usLocale KeepZeros 10.10 --> "10.10"
-```
+1. `Min Int` - will show at least a certain amount of decimal digits, adding trailing zeros if needed
+2. `Max Int` - will show up to a certain amount of decimal digits, discarding trailing zeros if needed
````
```suggestion
* `Max Int` shows up to a certain amount of decimal digits, discarding trailing zeros if needed.
```

cuducos commented on README.md:
````diff
@@ -32,21 +32,13 @@ format usLocale -pi --> "−3.14"
-```
+1. `Min Int` - will show at least a certain amount of decimal digits, adding trailing zeros if needed
+2. `Max Int` - will show up to a certain amount of decimal digits, discarding trailing zeros if needed
+3. `Exact Int` - will show an exact number of decimal digits, adding trailing zeros if needed
````
```suggestion
* `Exact Int` shows an exact number of decimal digits, adding trailing zeros if needed.
```

cuducos commented on README.md:
````diff
@@ -32,21 +32,13 @@ format usLocale -pi --> "−3.14"
 ```
 
-## Humanize
+### Decimals
````
```suggestion
### The `Decimals` strategy type
```

cuducos commented on src/FormatNumber.elm:
```diff
@@ -28,47 +21,46 @@ Just convert them to `Float` before passing them to `format`:
 {-| Format a float number as a pretty string:
 
-    import FormatNumber.Locales exposing (Locale, frenchLocale, spanishLocale, usLocale)
+    import FormatNumber.Locales exposing (Locale, frenchLocale, spanishLocale, usLocale, Decimals(..))
```
```suggestion
    import FormatNumber.Locales exposing (Decimals(..), Locale, frenchLocale, spanishLocale, usLocale)
```

cuducos commented on src/FormatNumber/Locales.elm:
```diff
@@ -1,6 +1,7 @@
 module FormatNumber.Locales exposing
     ( Locale
     , frenchLocale, spanishLocale, usLocale
+    , Decimals(..), base, getDecimals
```
I like to be consistent in the order in `module` and `import`. Unfortunately, I cannot suggest a multi-line change on GitHub, but let's write this import as `module FormatNumber.Locales exposing (Decimals(..), Locale, base, frenchLocale, getDecimals, spanishLocale, usLocale)` and let Elm Format does their magic?

cuducos commented on src/Parser.elm:
```diff
@@ -100,25 +99,25 @@ splitThousands integers =
 {-| Given a `Locale` parses a `Float` into a `FormattedNumber`:
 
-    import FormatNumber.Locales exposing (usLocale)
+    import FormatNumber.Locales exposing (usLocale, Decimals(..))
```
```suggestion
    import FormatNumber.Locales exposing (Decimals(..), usLocale)
```

cuducos reviewed (approved):
Hi you both, really really good! This API is looking so good ❤️ 

Thank you very much for this contribution ✨ 

I added minor edits, mostly related to style. Feel free to change them or not. I can also go over them once I merge it here. 

I'm not on a computer with the Elm toolchain, so I'll come back to this code over the weekend:
* One thing I might try is to think about the API for `getDecimals` (maybe try to expose it, maybe put it somewhere else);
* Another thing I might try is to reduce de cyclomatic complexity (does it make sense to say that to functional languages like Elm?) and enhance the readability of `Parser.elm` (especially lines 231-240 and 256-284) ; probably just try to extract the logic to auxiliary functions. 

Anyway, I'm just thinking out loud at this point. The contribution itself is solid and means a great enhancement for this library! So it's up to you know: I can merge and move on with these really minor things, but feel free to continue working on that! Just let me know how you prefer these next steps to be.

A huge thank you for your both 👍

malidukica committed:
Apply suggestions from code review

Co-Authored-By: Eduardo Cuducos <cuducos@users.noreply.github.com>

malidukica committed:
Update exposing statement

malidukica commented:
Hi! The suggestions accepted and comitted to the branch.
